lib/engine/game/g_18_fr/game.rb

```diff
@@ -0,0 +1,225 @@
+            corporation_sizes: [2, 5],
+          },
+          {
+            name: 'Brown',
+            on: '5',
+            train_limit: 3,
+            tiles: %i[yellow green brown],
+            operating_rounds: 2,
+            corporation_sizes: [2, 5],
+          },
+          {
+            name: 'Gray',
+            on: '6*D',
+            train_limit: 3,
+            tiles: %i[yellow green brown],
+            operating_rounds: 2,
+            corporation_sizes: [2, 5],
+          },
+        ].freeze
+
+        TRAINS = [{ name: '2', distance: 2, price: 100, rusts_on: '3P', num: 40 },
+                  { name: '2+', distance: 2, price: 100, obsolete_on: '3P', num: 4 },
+                  {
+                    name: '3+',
+                    distance: 3,
+                    price: 300,
+                    obsolete_on: '6*D',
+                    num: 12,
+                    events: [{ 'type' => 'two_tile_lays' }],
+                  },
+                  {
+                    name: '3P',
+                    distance: 3,
+                    price: 400,
+                    num: 1,
+                    events: [{ 'type' => 'free_ports' }],
+                  },
+                  { name: '2P', distance: 2, price: 300, num: 5 },
+                  { name: '5', distance: 5, price: 600, num: 6 },
+                  { name: '6*D', distance: 6, price: 800, num: 30 },
+                  { name: '2P*', distance: 2, price: 200, num: 1 }].freeze
+
+        EVENTS_TEXT = Base::EVENTS_TEXT.merge(
+          'two_tile_lays' => ['Corporations may lay two tiles for 20 F. One may be upgrade'],
+          'free_ports' => ['Ports no longer count towards train length']
+        ).freeze
+
+        ONE_YELLOW_TILE_LAY = [{ lay: true, upgrade: false }].freeze
+        TWO_TILE_LAYS = [
+          { lay: true, upgrade: true },
+          { lay: true, upgrade: :not_if_upgraded, cost: 20, cannot_reuse_same_hex: true },
+        ].freeze
+
+        B_HEX_NAMES = %w[F3].freeze
+        YELLOW_B_TILE_NAME = 'FRBY'
+        GREEN_B_TILE_NAME = 'FRBG'
+
+        def setup
+          @extra_tile_lay = false
+        end
+
+        def init_round
+          # skipping the initial auction for now
+          @log << "-- #{round_description('Stock', 1)} --"
+          @round_counter = 1
+          stock_round
+        end
+
+        def stock_round
+          close_bank_shorts
+          @interest_fixed = nil
+
+          G1817::Round::Stock.new(self, [
+            Engine::Step::DiscardTrain,
+            Engine::Step::HomeToken,
+            G18FR::Step::BuySellParShares,
+          ])
+        end
+
+        def operating_round(round_num)
+          @interest_fixed = nil
+          @interest_fixed = interest_rate
+          # Revaluate if private companies are owned by corps with trains
+          @companies.each do |company|
+            next unless company.owner
+
+            abilities(company, :revenue_change, time: 'has_train') do |ability|
```

Comment:
`Game.abilities` can return nil, a single ability object, or an array of ability objects, depending on how many matching abilities are found. This code is going to fail unless the company has multiple revenue change abilities. And if a company has multiple revenue change abilities, should you be summing the revenues?